File: cardano-rpc/quickstart/python/send_lovelace.py
import time
from fractions import Fraction
from pathlib import Path
from typing import List

import grpc
from pycardano import (
    Address,
    ChainContext,
    Network,
    PaymentSigningKey,
    ProtocolParameters,
    TransactionBuilder,
    TransactionInput,
    TransactionOutput,
    UTxO,
    Value,
)
from pycardano.hash import TransactionId
from utxorpc_spec.utxorpc.v1beta.cardano.cardano_pb2 import (
    AddressPattern,
    TxOutputPattern,
)
from utxorpc_spec.utxorpc.v1beta.query.query_pb2 import (
    AnyUtxoPattern,
    ReadParamsRequest,
    SearchUtxosRequest,
    UtxoPredicate,
)
from utxorpc_spec.utxorpc.v1beta.query.query_pb2_grpc import QueryServiceStub
from utxorpc_spec.utxorpc.v1beta.submit.submit_pb2 import AnyChainTx, SubmitTxRequest
from utxorpc_spec.utxorpc.v1beta.submit.submit_pb2_grpc import SubmitServiceStub

RPC_URL = "localhost:50051"
CLUSTER_DIR = "/tmp/demo-cluster"
LOVELACE_TO_SEND = 5_000_000  # 5 ADA (1 ADA = 1,000,000 lovelace)


class CardanoRpcChainContext(ChainContext):
    """A PyCardano ChainContext backed directly by cardano-rpc's v1beta gRPC services.

    Covers only what building, signing, and submitting a plain payment needs:
    UTxO lookup, protocol parameters, and submission.
    It is demo scaffolding, not a general-purpose provider: it does not surface
    datums, inline datums, reference scripts, or script cost models, and every
    BigInt field is read through its `int` branch, which testnet-scale values
    always fit.
    """

    def __init__(self, target: str):
        channel = grpc.insecure_channel(target)
        self._query = QueryServiceStub(channel)
        self._submit = SubmitServiceStub(channel)

    @property
    def network(self) -> Network:
        return Network.TESTNET

    @property
    def protocol_param(self) -> ProtocolParameters:
        pparams = self._query.ReadParams(ReadParamsRequest(), timeout=5).values.cardano

        def fraction(r) -> Fraction:
            return Fraction(r.numerator, r.denominator)

        return ProtocolParameters(
            min_fee_constant=pparams.min_fee_constant.int,
            min_fee_coefficient=pparams.min_fee_coefficient.int,
            max_block_size=pparams.max_block_body_size,
            max_tx_size=pparams.max_tx_size,
            max_block_header_size=pparams.max_block_header_size,
            key_deposit=pparams.stake_key_deposit.int,
            pool_deposit=pparams.pool_deposit.int,
            pool_influence=fraction(pparams.pool_influence),
            monetary_expansion=fraction(pparams.monetary_expansion),
            treasury_expansion=fraction(pparams.treasury_expansion),
            # Byron-era fields with no cardano-rpc equivalent; unused by a
            # plain Conway-era payment.
            decentralization_param=Fraction(0),
            extra_entropy="",
            min_utxo=0,
            coins_per_utxo_word=0,
            protocol_major_version=pparams.protocol_version.major,
            protocol_minor_version=pparams.protocol_version.minor,
            min_pool_cost=pparams.min_pool_cost.int,
            price_mem=fraction(pparams.prices.memory),
            price_step=fraction(pparams.prices.steps),
            max_tx_ex_mem=pparams.max_execution_units_per_transaction.memory,
            max_tx_ex_steps=pparams.max_execution_units_per_transaction.steps,
            max_block_ex_mem=pparams.max_execution_units_per_block.memory,
            max_block_ex_steps=pparams.max_execution_units_per_block.steps,
            max_val_size=pparams.max_value_size,
            collateral_percent=pparams.collateral_percentage,
            max_collateral_inputs=pparams.max_collateral_inputs,
            coins_per_utxo_byte=pparams.coins_per_utxo_byte.int,
            cost_models={},  # not needed to build a plain (non-script) payment
        )

    def _utxos(self, address: str) -> List[UTxO]:
        address_bytes = bytes(Address.decode(address).to_primitive())
        predicate = UtxoPredicate(
            match=AnyUtxoPattern(
                cardano=TxOutputPattern(
                    address=AddressPattern(exact_address=address_bytes)
                )
            )
        )
        response = self._query.SearchUtxos(
            SearchUtxosRequest(predicate=predicate), timeout=5
        )
        utxos = []
        for item in response.items:
            cardano = item.cardano
            tx_in = TransactionInput(
                TransactionId(item.txo_ref.hash), item.txo_ref.index
            )
            tx_out = TransactionOutput(
                Address.from_primitive(cardano.address), Value(cardano.coin.int)
            )
            utxos.append(UTxO(tx_in, tx_out))
        return utxos

    def submit_tx_cbor(self, cbor):
        if isinstance(cbor, str):
            cbor = bytes.fromhex(cbor)
        response = self._submit.SubmitTx(
            SubmitTxRequest(tx=AnyChainTx(raw=cbor)), timeout=5
        )
        return response.ref.hex()


def main() -> None:
    cluster_dir = Path(CLUSTER_DIR)
    context = CardanoRpcChainContext(RPC_URL)

    # cardano-testnet's genesis UTxO signing key is a TextEnvelope with type
    # "GenesisUTxOSigningKey_ed25519"; PaymentSigningKey.from_json() only
    # validates the type when asked to, so it loads this key as-is.
    signing_key = PaymentSigningKey.from_json(
        (cluster_dir / "utxo-keys/utxo1/utxo.skey").read_text()
    )
    sender_address = Address.load(str(cluster_dir / "utxo-keys/utxo1/utxo.addr"))
    recipient_address = Address.load(str(cluster_dir / "utxo-keys/utxo2/utxo.addr"))

    utxos = context.utxos(sender_address)
    print(f"Sender address:    {sender_address}")
    print(f"Recipient address: {recipient_address}")
    print(f"Spendable UTxOs:   {len(utxos)}")

    builder = TransactionBuilder(context)
    for utxo in utxos:
        builder.add_input(utxo)
    builder.add_output(TransactionOutput(recipient_address, Value(LOVELACE_TO_SEND)))

    signed_tx = builder.build_and_sign([signing_key], change_address=sender_address)
    tx_hash = context.submit_tx(signed_tx)
    print(f"Submitted tx: {tx_hash}")

    # Confirm by polling the recipient's UTxOs for the new output.
    deadline = time.monotonic() + 60
    while time.monotonic() < deadline:
        for utxo in context.utxos(recipient_address):
            if str(utxo.input.transaction_id) == tx_hash:
                print(
                    f"Confirmed: {utxo.output.amount.coin} lovelace landed at "
                    f"{recipient_address} ({tx_hash}#{utxo.input.index})"
                )
                return
        time.sleep(3)
    raise TimeoutError(
        "Timed out waiting for the new UTxO to appear at the recipient address"
    )


if __name__ == "__main__":
    main()

File: cardano-rpc/quickstart/haskell/app/SendLovelace.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE NumericUnderscores #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TypeApplications #-}

module Main (main) where

import Cardano.Api
import qualified Cardano.Api.Experimental as Exp
import qualified Cardano.Api.Experimental.Tx as Exp
import qualified Cardano.Api.Ledger as L
import Cardano.Rpc.Client (defMessage, (&), (.~), (^.))
import qualified Cardano.Rpc.Client as Rpc
import qualified Cardano.Rpc.Proto.Api.UtxoRpc.Query as Query
import qualified Cardano.Rpc.Proto.Api.UtxoRpc.Submit as Submit

import qualified Cardano.Ledger.Coin as L
import qualified Cardano.Ledger.Conway.Core as L
import Cardano.Slotting.EpochInfo (fixedEpochInfo)
import Cardano.Slotting.Time (mkSlotLength)

import Control.Concurrent (threadDelay)
import Control.Monad (when)
import Control.Monad.Catch (MonadThrow (..))
import qualified Data.ByteString as BS
import qualified Data.ByteString.Base16 as Base16
import Data.Default.Class (def)
import Data.Maybe (listToMaybe)
import qualified Data.Text as T (show, strip, unpack)
import qualified Data.Text.Encoding as T (decodeUtf8)
import qualified Data.Text.IO as T (putStrLn, readFile)
import Data.Time (UTCTime (..))
import Data.Time.Calendar (fromGregorian)
import GHC.Exts (IsList (..))

import UnliftIO.Exception (stringException, throwString)

-- | Address of a locally running node's gRPC endpoint, started with
-- @--enable-grpc-http@ (cardano-testnet), as set up in the quickstart README.
rpcAddress :: Rpc.Address
rpcAddress =
  Rpc.Address
    { Rpc.addressHost = "127.0.0.1"
    , Rpc.addressPort = 50_051
    , Rpc.addressAuthority = Nothing
    }

clusterDir :: FilePath
clusterDir = "/tmp/demo-cluster"

lovelaceToSend :: Integer
lovelaceToSend = 5_000_000 -- 5 ADA (1 ADA = 1,000,000 lovelace)

conwayEra :: Exp.Era ConwayEra
conwayEra = Exp.ConwayEra

main :: IO ()
main = do
  senderWitness <-
    orDie (docToString . prettyError)
      =<< readFileTextEnvelopeAnyOf
        [FromSomeType asType WitnessGenesisUTxOKey]
        (File $ clusterDir <> "/utxo-keys/utxo1/utxo.skey")

  senderAddressText <- T.strip <$> T.readFile (clusterDir <> "/utxo-keys/utxo1/utxo.addr")
  recipientAddressText <- T.strip <$> T.readFile (clusterDir <> "/utxo-keys/utxo2/utxo.addr")
  senderAddress <- parseAddress conwayEra senderAddressText
  recipientAddress <- parseAddress conwayEra recipientAddressText

  T.putStrLn $ "Sender address:    " <> senderAddressText
  T.putStrLn $ "Recipient address: " <> recipientAddressText

  Rpc.withConnection def (Rpc.ServerInsecure rpcAddress) $ \conn -> do
    utxoItems <- searchUtxos conwayEra conn senderAddress
    T.putStrLn $ "Spendable UTxOs:   " <> T.show (length utxoItems)

    pparams <- readProtocolParams conwayEra conn

    utxoTxIns <- traverse (itemToTxIn . (^. Query.txoRef)) utxoItems
    itemCoins <- traverse itemCoin utxoItems
    let totalInputLovelace = sum itemCoins
        recipientTxOut = Exp.TxOut $ mkLedgerTxOut conwayEra recipientAddress lovelaceToSend
        ledgerUtxo =
          L.UTxO $
            fromList
              [ (toShelleyTxIn txIn, mkLedgerTxOut conwayEra senderAddress coin)
              | (txIn, coin) <- zip utxoTxIns itemCoins
              ]

    when (totalInputLovelace <= lovelaceToSend) $
      throwString "Sender's UTxOs do not cover 5 ADA plus fees"

    signedTx <-
      buildAndSignTx conwayEra senderWitness ledgerUtxo utxoTxIns recipientTxOut senderAddress pparams
    let rawTxBytes = serialiseToRawBytes signedTx

    submitResponse <-
      Rpc.nonStreaming conn (Rpc.rpc @(Rpc.Protobuf Submit.SubmitService "submitTx")) $
        defMessage & Submit.tx .~ (defMessage & Submit.raw .~ rawTxBytes)
    let submittedTxHash = submitResponse ^. Submit.ref
    T.putStrLn $ "Submitted tx: " <> hexEncode submittedTxHash

    confirmDelivery conwayEra conn recipientAddress recipientAddressText submittedTxHash

-- | Build a transaction spending @utxoTxIns@ into @recipientTxOut@, balanced
-- and signed by cardano-api's 'Exp.makeTransactionBodyAutoBalance': it works
-- out the minimum fee and the change output back to the sender itself, given
-- the ledger 'L.PParams' fetched by 'readProtocolParams'.
buildAndSignTx
  :: MonadThrow m
  => Exp.Era era
  -- ^ The era to build the transaction in.
  -> ShelleyWitnessSigningKey
  -- ^ Signs the transaction; must match the change address below.
  -> L.UTxO (Exp.LedgerEra era)
  -- ^ All inputs being spent, needed to work out their total value for balancing.
  -> [TxIn]
  -- ^ The transaction inputs to spend.
  -> Exp.TxOut (Exp.LedgerEra era)
  -- ^ The payment output to the recipient.
  -> AddressInEra era
  -- ^ Change address; the leftover balance is sent back here.
  -> L.PParams (Exp.LedgerEra era)
  -- ^ Protocol parameters; only the fields 'readProtocolParams' maps are read.
  -> m (Exp.SignedTx era)
buildAndSignTx era senderWitness ledgerUtxo utxoTxIns recipientTxOut senderAddress pparams = do
  let content =
        Exp.defaultTxBodyContent
          & Exp.setTxIns [(txIn, Exp.AnyKeyWitnessPlaceholder) | txIn <- utxoTxIns]
          & Exp.setTxOuts [recipientTxOut]

  (unsignedTx, _balancedContent) <-
    either (throwM . stringException . docToString . prettyError) pure $
      Exp.obtainCommonConstraints era $
        Exp.makeTransactionBodyAutoBalance
          offlineSystemStart
          offlineEpochInfo
          pparams
          mempty
          mempty
          ledgerUtxo
          content
          senderAddress
          Nothing

  let keyWitness = Exp.makeKeyWitness era unsignedTx senderWitness
  pure $ Exp.signTx era [] [keyWitness] unsignedTx

-- | Only used to convert slots to POSIX time for Plutus scripts; this
-- transaction has none, so any fixed epoch shape here is safe.
offlineSystemStart :: SystemStart
offlineSystemStart = SystemStart $ UTCTime (fromGregorian 2021 9 1) 0

-- | Only used by the auto-balancer to translate slots to wall-clock time for
-- Plutus scripts; this transaction has none, so a fixed shape avoids an extra RPC call.
offlineEpochInfo :: LedgerEpochInfo
offlineEpochInfo = LedgerEpochInfo $ fixedEpochInfo (EpochSize 1) (mkSlotLength 1)

mkLedgerTxOut :: Exp.Era era -> AddressInEra era -> Integer -> L.TxOut (Exp.LedgerEra era)
mkLedgerTxOut era address lovelace =
  Exp.obtainCommonConstraints era $
    L.mkBasicTxOut (toShelleyAddr address) (L.inject $ L.Coin lovelace)

-- | Poll the recipient's UTxOs until one matches the submitted transaction
-- hash, for up to a minute.
confirmDelivery
  :: Exp.Era era
  -- ^ The era the recipient address and submitted transaction belong to.
  -> Rpc.Connection
  -- ^ Open connection to cardano-rpc.
  -> AddressInEra era
  -- ^ Address to poll for the new UTxO.
  -> Text
  -- ^ The same address, pre-rendered for the confirmation message.
  -> BS.ByteString
  -- ^ Hash of the submitted transaction, to match against incoming UTxOs.
  -> IO ()
confirmDelivery era conn recipientAddress recipientAddressText submittedTxHash = go (20 :: Int)
 where
  go attemptsLeft
    | attemptsLeft <= 0 =
        throwString "Timed out waiting for the new UTxO to appear at the recipient address"
    | otherwise = do
        utxoItems <- searchUtxos era conn recipientAddress
        case listToMaybe [item | item <- utxoItems, item ^. Query.txoRef . Query.hash == submittedTxHash] of
          Just item -> do
            coin <- itemCoin item
            T.putStrLn $
              "Confirmed: "
                <> T.show coin
                <> " lovelace landed at "
                <> recipientAddressText
                <> " ("
                <> hexEncode submittedTxHash
                <> "#"
                <> T.show (item ^. Query.txoRef . Query.index)
                <> ")"
          Nothing -> do
            threadDelay 3_000_000
            go (attemptsLeft - 1)

searchUtxos :: Exp.Era era -> Rpc.Connection -> AddressInEra era -> IO [Rpc.Proto Query.AnyUtxoData]
searchUtxos era conn address = do
  response <-
    Rpc.nonStreaming conn (Rpc.rpc @(Rpc.Protobuf Query.QueryService "searchUtxos")) $
      defMessage & Query.predicate .~ exactAddressPredicate era address
  pure $ response ^. Query.items

-- | Fetch the current ledger protocol parameters via utxorpc's 'QueryService'
-- @readParams@ method (the same channel every other language example uses),
-- then map only the fields this script-free, single-payment transaction's
-- balancing consumes: the fee formula (per-byte and fixed) and the coins-per-
-- UTxO-byte rate the balancer needs for the change output's minimum UTxO.
-- 'Exp.makeTransactionBodyAutoBalance' never reads the maximum transaction
-- size at build time, so it is left unmapped; everything else (cost models,
-- deposits, governance thresholds, ...) stays at 'L.emptyPParams'' zero or
-- empty default, which this no-script, no-certificate transaction never touches.
readProtocolParams :: Exp.Era era -> Rpc.Connection -> IO (L.PParams (Exp.LedgerEra era))
readProtocolParams era conn = do
  response <-
    Rpc.nonStreaming conn (Rpc.rpc @(Rpc.Protobuf Query.QueryService "readParams")) defMessage
  let pparams = response ^. Query.values . Query.cardano

  minFeeCoefficient <- bigIntToInteger (pparams ^. Query.minFeeCoefficient)
  minFeeConstant <- bigIntToInteger (pparams ^. Query.minFeeConstant)
  coinsPerUtxoByte <- bigIntToInteger (pparams ^. Query.coinsPerUtxoByte)
  minFeeCoefficientCompact <- toCompactCoin "minFeeCoefficient" minFeeCoefficient
  coinsPerUtxoByteCompact <- toCompactCoin "coinsPerUtxoByte" coinsPerUtxoByte

  pure $
    Exp.obtainCommonConstraints era $
      L.emptyPParams
        & L.ppTxFeePerByteL .~ L.CoinPerByte minFeeCoefficientCompact
        & L.ppTxFeeFixedL .~ L.Coin minFeeConstant
        & L.ppCoinsPerUTxOByteL .~ L.CoinPerByte coinsPerUtxoByteCompact

-- | Pack a non-negative lovelace amount into the ledger's compact coin
-- representation, as 'L.CoinPerByte' requires.
toCompactCoin
  :: String
  -- ^ Field name, used only to name the offending value in the error.
  -> Integer
  -> IO (L.CompactForm L.Coin)
toCompactCoin fieldName value =
  maybe
    (throwString $ "Protocol parameter " <> fieldName <> " does not fit a compact coin: " <> show value)
    (pure . L.CompactCoin)
    (L.integerToWord64 value)

itemCoin :: Rpc.Proto Query.AnyUtxoData -> IO Integer
itemCoin item = bigIntToInteger (item ^. Query.cardano . Query.coin)

itemToTxIn :: MonadThrow m => Rpc.Proto Query.TxoRef -> m TxIn
itemToTxIn txoRef =
  case deserialiseFromRawBytes AsTxId (txoRef ^. Query.hash) of
    Left err -> throwM . stringException $ "Malformed txo ref hash from cardano-rpc: " <> show err
    Right utxoTxId -> pure $ TxIn utxoTxId (TxIx (fromIntegral (txoRef ^. Query.index)))

-- | Decode a UTxO RPC 'Query.BigInt': its value is stored in one of three
-- oneof variants depending on magnitude and sign.
bigIntToInteger :: Rpc.Proto Query.BigInt -> IO Integer
bigIntToInteger bigInt
  | Just int <- bigInt ^. Query.maybe'int = pure $ toInteger int
  | Just bytes <- bigInt ^. Query.maybe'bigUInt = toInteger <$> naturalFromRawBytes bytes
  | Just bytes <- bigInt ^. Query.maybe'bigNInt = do
      n <- naturalFromRawBytes bytes
      pure $ negate (toInteger n) - 1
  | otherwise = pure 0
 where
  naturalFromRawBytes bytes =
    either (throwString . show) pure $ deserialiseFromRawBytes AsNatural bytes

-- | Match a UTxO exactly by its address, mirroring cardano-rpc's own
-- (internal) 'exactAddressPredicate'.
exactAddressPredicate :: Exp.Era era -> AddressInEra era -> Rpc.Proto Query.UtxoPredicate
exactAddressPredicate era address =
  Exp.obtainCommonConstraints era $
    defMessage
      & Query.match
        .~ ( defMessage
               & Query.cardano
                 .~ (defMessage & Query.address .~ (defMessage & Query.exactAddress .~ serialiseToRawBytes address))
           )

parseAddress :: Exp.Era era -> Text -> IO (AddressInEra era)
parseAddress era addressText =
  Exp.obtainCommonConstraints era $
    orDie id $
      maybe (Left $ "Could not parse address: " <> T.unpack addressText) Right $
        deserialiseAddress (AsAddressInEra asType) addressText

hexEncode :: BS.ByteString -> Text
hexEncode = T.decodeUtf8 . Base16.encode

orDie :: (e -> String) -> Either e a -> IO a
orDie render = either (throwString . render) pure
